manifests/timer.pp

```diff
@@ -0,0 +1,102 @@
+# Create a timer and optionally a service unit
+# to execute with the timer unit.
+#
+# @api public
+#
+# @see systemd.timer
+#
+# @param name [Pattern['^.+\.timer$]]
+#   The target of the timer unit to create
+#
+# @param path
+#   The main systemd configuration path
+#
+# @param timer_content
+#   The full content of the timer unit file
+#
+#   * Mutually exclusive with ``$timer_source``
+#
+# @param timer_source
+#   The ``File`` resource compatible ``source``
+#
+#   * Mutually exclusive with ``$timer_content``
+#
+# @param service_content
+#   The full content of the service unit file
+#
+#   * Mutually exclusive with ``$timer_source``
+#
+# @param service_source
+#   The ``File`` resource compatible ``source``
+#
+#   * Mutually exclusive with ``$service_content``
+#
+# @param owner
+#   The owner to set on the dropin file
+#
+# @param group
+#   The group to set on the dropin file
+#
+# @param mode
+#   The mode to set on the dropin file
+#
+# @param show_diff
+#   Whether to show the diff when updating dropin file
+#
+# @param service_unit
+#   If set then the service_unit will have this name.
+#   If not set the service unit has the same name
+#   as the timer unit with s/.timer/.service/
+#
+define systemd::timer (
+  Enum['present', 'absent', 'file']        $ensure = 'present',
+  Stdlib::Absolutepath                     $path = '/etc/systemd/system',
+  Optional[String[1]]                      $timer_content = undef,
+  Optional[String[1]]                      $timer_source = undef,
+  Optional[String[1]]                      $service_content = undef,
+  Optional[String[1]]                      $service_source  = undef,
+  String[1]                                $owner = 'root',
+  String[1]                                $group = 'root',
+  Stdlib::Filemode                         $mode = '0444',
+  Optional[Systemd::Unit]                  $service_unit = undef,
+  Boolean                                  $show_diff = true,
+  Optional[Variant[Boolean, Enum['mask']]] $enable = undef,
+  Optional[Boolean]                        $active = undef,
+) {
+  assert_type(Pattern['^.+\.timer$'],$name)
+
+  if $service_unit {
+    $_service_unit = $service_unit
+  } else {
+    $_service_unit = "${basename($name,'.timer')}.service"
+  }
+
+  contain systemd::systemctl::daemon_reload
```

Comment:
Doesn't the unit file handle this already?

Reply:
Not quite in my experience the order is not correct and deamon-reload often happens after the service is 
managed, looking a bit harder now. `unit_file` contains:

```puppet
   include systemd    # which contains  daemon_reload'
     file { "${path}/${name}":
     ensure    => $_ensure,
     show_diff => $show_diff,
     notify    => Class['systemd::systemctl::daemon_reload'],
  }
service { $name:
  ensure    => $active,
  subscribe => File["${path}/${name}"],
  require   => Class['systemd::systemctl::daemon_reload'],
}

```
there is nothing in there that forces the exec of daemon-reload to happen before the service is managed.

Probably if unit_file itself contained `systemd::systemctl::daemon_reload`  or prossibly `contain systemd` `it might work.

Reply:
> there is nothing in there that forces the exec of daemon-reload to happen before the service is managed.

The `service` has a `require => Class['systemd::systemctl::daemon_reload']`. Doesn't that take care of it?

Reply:
No it does not. 

The class needs to be contained where the require is done. 
With include class you are saying I require that class be included before.
With contain class you are saying I require that class be included and do everything inside that class.

`require -> Class['abc']`  on its own is almost always not what you want with out a contain (or anchor)

Reply:
Another MR s/include/contain/ problably thing to do. Can do in a week or two ....

Reply:
That's not my understanding of how contain works. AFAIK `contain x` in `class a` it means that if I `require => Class['a']` I'll be sure that `class x` is executed before. However, the service still requires it and within the catalog it should be ordered correctly.

That said, with `include` and just `notify => Class['daemon_reload']` you can't be sure with just `require => Systemd::Unit_file['...']` that it has reloaded.